File: src/transposition/RailFence.kt
package transposition

import java.lang.IndexOutOfBoundsException

object RailFence {

    fun encipher(plainText: String): String {
        val upperCaseText = plainText.toUpperCase()
        val inputStream = upperCaseText.toCharArray()
        val outputStream1 = mutableListOf<Char>()
        val outputStream2 = mutableListOf<Char>()
        inputStream.forEachIndexed { index, c ->
            if ((index % 2) == 0) outputStream1.add(c)
            else outputStream2.add(c)
        }
        val outputStream = outputStream1 + outputStream2
        return outputStream.joinToString("")
    }

    fun decipher(cipherText: String): String {
        val cipherLength = cipherText.length
        val upperCaseText = cipherText.toUpperCase()
        val inputStream = upperCaseText.toCharArray()
        val outputStream = mutableListOf<Char>()
        val midPoint = when (cipherLength % 2) {
            0 -> (cipherLength) / 2
            else -> (cipherLength + 1) / 2
        }
        for (index in 0 until midPoint) {
            try {
                outputStream.add(inputStream[index])
                outputStream.add(inputStream[index + midPoint])
            } catch (e: IndexOutOfBoundsException) {
                // Ignore
            }
        }
        return outputStream.joinToString("")
    }
}

File: C/transposition/railfence.c
#include "transposition.h"
#include "railfence.h"

void encrypt_railfence(char* plain_text, char* cipher_text) {
    int message_length = (int) strlen(plain_text);
    int mid_point = (message_length % 2 == 0)?
            message_length / 2 : (message_length + 1) / 2;
    int i = 0;
    while (plain_text[i] != 0) {
        if (i % 2 == 0) {
            cipher_text[i / 2] = (char) toupper(plain_text[i]);
        } else {
            cipher_text[i / 2 + mid_point] = (char) toupper(plain_text[i]);
        }
        i++;
    }
    cipher_text[i] = 0;
}

void decrypt_railfence(char* cipher_text, char* plain_text) {
    int message_length = (int) strlen(cipher_text);
    int mid_point = (message_length % 2 == 0)?
                    message_length / 2 : (message_length + 1) / 2;
    int i = 0;
    while (cipher_text[i] != 0) {
        if (i < mid_point) {
            plain_text[i * 2] = (char) toupper(cipher_text[i]);
        } else {
            plain_text[i * 2 + 1] = (char) toupper(cipher_text[i]);
        }
        i++;
    }
    plain_text[i] = 0;
}